File: haskell/day02/app/Main.hs
{-# LANGUAGE DerivingStrategies #-}

module Main where

import Data.Maybe (fromJust)

type Horizonal = Int

type Depth = Int

type Aim = Int

data Move = Forward Int | Up Int | Down Int
  deriving stock (Eq, Show)

makeMoveConstructorFromString :: String -> Maybe (Int -> Move)
makeMoveConstructorFromString "forward" = Just Forward
makeMoveConstructorFromString "down" = Just Down
makeMoveConstructorFromString "up" = Just Up
makeMoveConstructorFromString _ = Nothing

parseInputFile :: FilePath -> IO (Maybe [Move])
parseInputFile path = do
  content <- readFile path
  let substrings :: [(String, String)]
      substrings = map ((\ws -> (ws !! 0, ws !! 1)) . words) (lines content)
  return $ traverse (\(s, c) -> makeMoveConstructorFromString s <*> pure (read c)) substrings

calculateHorizontalPosition :: [Move] -> Horizonal
calculateHorizontalPosition [] = 0
calculateHorizontalPosition (Forward x : ms) = x + calculateHorizontalPosition ms
calculateHorizontalPosition (_ : ms) = 0 + calculateHorizontalPosition ms

calculateDepthPosition :: [Move] -> Depth
calculateDepthPosition [] = 0
calculateDepthPosition (Forward _ : ms) = calculateDepthPosition ms
calculateDepthPosition (Up x : ms) = calculateDepthPosition ms - x
calculateDepthPosition (Down x : ms) = x + calculateDepthPosition ms

calculateWithAim :: (Horizonal, Depth, Aim) -> [Move] -> Int
calculateWithAim (horizontal, depth, _) [] = horizontal * depth
calculateWithAim (horizontal, depth, aim) (Forward x : ms) = calculateWithAim (horizontal + x, depth + aim * x, aim) ms
calculateWithAim (horizontal, depth, aim) (Up x : ms) = calculateWithAim (horizontal, depth, aim - x) ms
calculateWithAim (horizontal, depth, aim) (Down x : ms) = calculateWithAim (horizontal, depth, aim + x) ms

main :: IO ()
main = do
  moves <- parseInputFile "input.txt"
  let moves' = fromJust moves
  putStrLn $ "Answer 1: " <> show (calculateHorizontalPosition moves' * calculateDepthPosition moves')
  putStrLn $ "Answer 2: " <> show (calculateWithAim (0, 0, 0) moves')

File: rust/day02/src/main.rs
use std::fs;
use std::path::Path;
use std::str::FromStr;

#[derive(Debug)]
enum Move {
    Forward(i64),
    Down(i64),
    Up(i64),
}

type MoveConstructor = fn(i64) -> Move;

fn make_direction_from_string(s: &str) -> Option<MoveConstructor> 
{
    match s {
        "forward" => Some(Move::Forward),
        "down" => Some(Move::Down),
        "up" => Some(Move::Up),
        _ => None,
    }
}

fn parse_input_file<T: AsRef<Path>>(path: T) -> Vec<Move> {
    let content = fs::read_to_string(path).unwrap();
    content
        .lines()
        .map(|line| line.split(' '))
        .map(|mut substrings| (substrings.next().unwrap(), substrings.next().unwrap()))
        .map(|(move_str, count)| {
            make_direction_from_string(move_str).unwrap()(i64::from_str(count).unwrap())
        })
        .collect()
}

fn calculate_horizontal_position(moves: &[Move]) -> i64 {
    moves
        .iter()
        .map(|m| match m {
            Move::Forward(x) => *x,
            _ => 0,
        })
        .sum()
}

fn calculate_depth_position(moves: &[Move]) -> i64 {
    moves
        .iter()
        .map(|m| match m {
            Move::Up(x) => -x,
            Move::Down(x) => *x,
            _ => 0,
        })
        .sum()
}

fn calculate_with_aim(moves: &[Move]) -> i64 {
    let mut horizontal: i64 = 0;
    let mut depth: i64 = 0;
    let mut aim: i64 = 0;
    
    for m in moves {
        match m {
            Move::Forward(x) => { horizontal += x; depth += aim * x },
            Move::Down(x) => aim += x,
            Move::Up(x) => aim -= x,
        }
    }
    horizontal * depth
}

fn main() {
    let moves = parse_input_file("input.txt");
    println!(
        "Answer 1: {}",
        calculate_horizontal_position(&moves) * calculate_depth_position(&moves)
    );
    println!(
        "Answer 2: {}",
        calculate_with_aim(&moves),
    );
}
